File: py/moq-lite/moq_lite/publish.py
"""Producer wrappers — publish broadcasts and media tracks."""

from __future__ import annotations

from typing import TYPE_CHECKING

from moq_ffi import MoqBroadcastProducer, MoqMediaProducer, MoqRawGroupProducer, MoqRawProducer

if TYPE_CHECKING:
    from .subscribe import BroadcastConsumer, RawConsumer, RawGroupConsumer


class MediaProducer:
    """Wraps MoqMediaProducer with a cleaner interface."""

    def __init__(self, inner: MoqMediaProducer) -> None:
        self._inner = inner

    def write_frame(self, payload: bytes, timestamp_us: int) -> None:
        self._inner.write_frame(payload, timestamp_us)

    def finish(self) -> None:
        self._inner.finish()


class RawGroupProducer:
    """Writes frames into a single group on a raw track."""

    def __init__(self, inner: MoqRawGroupProducer) -> None:
        self._inner = inner

    @property
    def sequence(self) -> int:
        """The sequence number of this group within the track."""
        return self._inner.sequence()

    def consume(self) -> RawGroupConsumer:
        """Create a consumer that reads frames from this group."""
        from .subscribe import RawGroupConsumer

        return RawGroupConsumer(self._inner.consume())

    def write_frame(self, payload: bytes) -> None:
        self._inner.write_frame(payload)

    def finish(self) -> None:
        self._inner.finish()


class RawProducer:
    """Raw track producer — write arbitrary byte payloads with no codec required.

    Same pattern as moq-boy's status/command tracks.
    """

    def __init__(self, inner: MoqRawProducer) -> None:
        self._inner = inner

    def append_group(self) -> RawGroupProducer:
        """Start a new group; write frames into it, then finish()."""
        return RawGroupProducer(self._inner.append_group())

    def write_frame(self, payload: bytes) -> None:
        """Convenience: write a single-frame group in one call."""
        self._inner.write_frame(payload)

    def consume(self) -> RawConsumer:
        """Create a consumer that reads directly from this producer's track."""
        from .subscribe import RawConsumer

        return RawConsumer(self._inner.consume())

    def finish(self) -> None:
        self._inner.finish()


class BroadcastProducer:
    """Wraps MoqBroadcastProducer with a cleaner interface."""

    def __init__(self) -> None:
        self._inner = MoqBroadcastProducer()

    def publish_media(self, format: str, init: bytes) -> MediaProducer:
        return MediaProducer(self._inner.publish_media(format, init))

    def publish_raw(self, name: str) -> RawProducer:
        """Create a raw track — send any bytes, no codec validation."""
        return RawProducer(self._inner.publish_raw(name))

    def consume(self) -> BroadcastConsumer:
        """Create a consumer that reads from this broadcast's tracks."""
        from .subscribe import BroadcastConsumer

        return BroadcastConsumer(self._inner.consume())

    def finish(self) -> None:
        self._inner.finish()

File: go/moq-lite/publish.go
package moq

import (
	"context"

	moqffi "github.com/moq-dev/moq/go/moq-ffi/moq"
)

// MediaProducer publishes media frames on a media track.
type MediaProducer struct {
	inner *moqffi.MoqMediaProducer
}

func newMediaProducer(inner *moqffi.MoqMediaProducer) *MediaProducer {
	return &MediaProducer{inner: inner}
}

// Name returns the generated media track name.
func (m *MediaProducer) Name() (string, error) { return m.inner.Name() }

// Used blocks until this track has at least one active subscriber.
func (m *MediaProducer) Used(ctx context.Context) error {
	_, err := callCtx(ctx, func() {}, func() (struct{}, error) {
		return struct{}{}, m.inner.Used()
	})
	return err
}

// Unused blocks until this track has no active subscribers.
func (m *MediaProducer) Unused(ctx context.Context) error {
	_, err := callCtx(ctx, func() {}, func() (struct{}, error) {
		return struct{}{}, m.inner.Unused()
	})
	return err
}

func (m *MediaProducer) WriteFrame(payload []byte, timestampUs uint64) error {
	return m.inner.WriteFrame(payload, timestampUs)
}

func (m *MediaProducer) Finish() error { return m.inner.Finish() }
func (m *MediaProducer) Close()        { m.inner.Destroy() }

// GroupProducer writes frames into a single group on a track.
type GroupProducer struct {
	inner *moqffi.MoqGroupProducer
}

func newGroupProducer(inner *moqffi.MoqGroupProducer) *GroupProducer {
	return &GroupProducer{inner: inner}
}

func (g *GroupProducer) Sequence() uint64          { return g.inner.Sequence() }
func (g *GroupProducer) WriteFrame(p []byte) error { return g.inner.WriteFrame(p) }
func (g *GroupProducer) Finish() error             { return g.inner.Finish() }
func (g *GroupProducer) Close()                    { g.inner.Destroy() }

// Consume returns a consumer that reads frames from this group.
func (g *GroupProducer) Consume() (*GroupConsumer, error) {
	c, err := g.inner.Consume()
	if err != nil {
		return nil, err
	}
	return newGroupConsumer(c), nil
}

// TrackProducer publishes arbitrary byte frames on a non-media track.
//
// Same pattern as moq-boy's status/command tracks.
type TrackProducer struct {
	inner *moqffi.MoqTrackProducer
}

func newTrackProducer(inner *moqffi.MoqTrackProducer) *TrackProducer {
	return &TrackProducer{inner: inner}
}

func (t *TrackProducer) Name() (string, error) { return t.inner.Name() }

// Used blocks until this track has at least one active subscriber.
func (t *TrackProducer) Used(ctx context.Context) error {
	_, err := callCtx(ctx, func() {}, func() (struct{}, error) {
		return struct{}{}, t.inner.Used()
	})
	return err
}

// Unused blocks until this track has no active subscribers.
func (t *TrackProducer) Unused(ctx context.Context) error {
	_, err := callCtx(ctx, func() {}, func() (struct{}, error) {
		return struct{}{}, t.inner.Unused()
	})
	return err
}

// AppendGroup opens a new group; write frames into it, then Finish.
func (t *TrackProducer) AppendGroup() (*GroupProducer, error) {
	g, err := t.inner.AppendGroup()
	if err != nil {
		return nil, err
	}
	return newGroupProducer(g), nil
}

// WriteFrame writes a single-frame group in one call.
func (t *TrackProducer) WriteFrame(p []byte) error { return t.inner.WriteFrame(p) }

// Consume returns a consumer that reads from this producer's track directly,
// without going through an origin/broadcast.
func (t *TrackProducer) Consume() (*TrackConsumer, error) {
	c, err := t.inner.Consume()
	if err != nil {
		return nil, err
	}
	return newTrackConsumer(c), nil
}

func (t *TrackProducer) Finish() error { return t.inner.Finish() }
func (t *TrackProducer) Close()        { t.inner.Destroy() }

// BroadcastProducer publishes tracks within a broadcast.
type BroadcastProducer struct {
	inner *moqffi.MoqBroadcastProducer
}

// NewBroadcastProducer creates a broadcast for publishing media and/or
// raw tracks. The broadcast does nothing until published to an origin.
func NewBroadcastProducer() (*BroadcastProducer, error) {
	b, err := moqffi.NewMoqBroadcastProducer()
	if err != nil {
		return nil, err
	}
	return &BroadcastProducer{inner: b}, nil
}

// PublishMedia opens a new media track. `format` is the codec identifier
// (e.g. "opus", "avc3"), `init` is the codec-specific init payload.
func (b *BroadcastProducer) PublishMedia(format string, init []byte) (*MediaProducer, error) {
	m, err := b.inner.PublishMedia(format, init)
	if err != nil {
		return nil, err
	}
	return newMediaProducer(m), nil
}

// PublishTrack opens a new raw byte-payload track, no codec validation.
func (b *BroadcastProducer) PublishTrack(name string) (*TrackProducer, error) {
	t, err := b.inner.PublishTrack(name)
	if err != nil {
		return nil, err
	}
	return newTrackProducer(t), nil
}

// Consume returns a consumer that reads from this broadcast's tracks
// directly, without going through an origin.
func (b *BroadcastProducer) Consume() (*BroadcastConsumer, error) {
	c, err := b.inner.Consume()
	if err != nil {
		return nil, err
	}
	return newBroadcastConsumer(c), nil
}

func (b *BroadcastProducer) Finish() error { return b.inner.Finish() }
func (b *BroadcastProducer) Close()        { b.inner.Destroy() }
